docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+
+If we decide to support watches spanning different resources in the future and
+we have an efficient indexing mechanisms, it should be relatively simple to unify
+the cache to be common for all the resources.
+
+Since we want the watch in apiserver to be optional for different resource
+types, this needs to be self-contained and hidden behind a well defined API.
+It seems that this API can be very simple - it should just expose a method
+for initializing the cache and another one for registering (and starting) a
+new watcher with a given filtering function (matcher).
+
+The rest of this section describes the concrete steps that need to be done
+to implement the proposal.
+
+1. Since we are initially targeting to have per-resource-type watch in
+apiserver, this should probably be incorporated into generic etcd registry:
+"pkg/registry/generic/etcd". Thus we should migrate all the resources to
+use it (it seems it's not the case now e.g. for services).
+
+2. We should define an API for the watch in apiserver (this will be purely
+internal api just within apiserver, so it should be possible to change it
+relatively easy in the future). Also, we should provide the initial
+implementation that will be just redirecting all the requests to etcd, similarly
+to what we are doing today. This simple implementation should be used to enable
+easy transition to the new mechanism later (api unification).
+
+3. The target implementation internally will have a single watch to etcd open,
+responsible for watching all the changes of objects of a given resource type.
+However, this watch can potentially expire at any time and reconnecting can
+return "too old resource version". In that case relisting is necessary.
+To avoid LIST requests coming from all the watcher at that time, we will
+introduce an additional etcd event type:
+[EtcdResync](../../pkg/tools/etcd_helper_watch.go#L36)
+
+  Whenever reslisting will be done to refresh the internal watch to etcd,
+  EtcdResync event will be send to all the watchers. It will contain the
+  full list of all the objects the watcher is interested in (appropriately
+  filtered) as the parameter of this watch event.
+  Thus, we need to create the EtcdResync event, extend watch.Interface and
+  its implementations to support it and handle those events appropriately
+  in places like
+  [Reflector](../../pkg/client/cache/reflector.go)
+
+4. Create a FIFO cache with a given capacity. This should be as simple as
+having an array an treating it as a cyclic one. Note that in the cache
+we should store two things: the object watched from etcd and its resource
+version (etcdIndex). Obviously resource versions will be increasing (if
+we treat the array as a cyclic one), but they are necessary for registering
+a new watcher that is interested in all the changes that happened after
+a given etcdIndex (resource version).
+
+  We should also support LIST operation from the beginning and the end of
```

Comment:
You have to support list somehow, otherwise clients can never start watching in the first place. You could pass lists through to etcd, but that won't work with an indexer, so I think you need to just keep everything in memory.

And again, I strongly suggest leaving the resync event to a second pass. I really think it's a premature optimization. Apiserver should basically always keep up.